core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -333,37 +346,55 @@ private static void logIfContextNarrowedTimeout(
 
   private void removeContextListenerAndCancelDeadlineFuture() {
     context.removeListener(cancellationListener);
-    ScheduledFuture<?> f = deadlineCancellationFuture;
+    ScheduledFuture<?> f = deadlineCancellationSendToServerFuture;
+    if (f != null) {
+      f.cancel(false);
+    }
+
+    f = deadlineCancellationNotifyApplicationFuture;
     if (f != null) {
       f.cancel(false);
     }
   }
 
-  private class DeadlineTimer implements Runnable {
+  private ScheduledFuture<?> startDeadlineSendCancelToServerTimer(
+      long remainingNanos, InsightBuilder insight) {
+    return deadlineCancellationExecutor.schedule(
+        new LogExceptionRunnable(new DeadlineSendCancelToServerTimer(remainingNanos, insight)),
```

Comment:
To make the code easier to follow and reduce the fragmentation, consider moving the whole `DeadlineSendCancelToServerTimer` into this method body (probably also eliminating the arguments to the timer constructor since it's now a closure).